### ISSUE TITLE
Location of the default lima socket is wrong and provider status "unknown"

### ISSUE THREAD

**Issue by afbjorklund** [containers/podman-desktop#2547]
### Bug description

It is looking for ~/.lima/podman-lima/sock/podman.sock, but the default location is ~/.lima/podman/sock/podman.sock

It would be nice if the `LIMA_HOME` environment variable was respected, as it can override the default `HOME` location.

### Operating system

Ubuntu 20.04

### Version

1.0.0

### Additional context

It should also look for the presence of the `limactl` program, and use that to determine the version to show in the GUI.

```console
$ limactl --version
limactl version 0.15.1
```

When the podman socket is missing, then the instance (virtual machine) is stopped. If it is present, it has been started:

```console
$ limactl list podman
NAME      STATUS     SSH                VMTYPE    ARCH      CPUS    MEMORY    DISK      DIR
podman    Running    127.0.0.1:40361    qemu      x86_64    4       4GiB      100GiB    ~/.lima/podman
```

**Comment by afbjorklund** [comment]
How to install and use with lima:

`brew install lima`

`limactl start template://podman`

`podman.lima version`

**Comment by afbjorklund** [comment]
Current workaround:

`limactl start --name=podman-lima template://podman`

`export LIMA_INSTANCE=podman-lima`

`podman.lima version`

**Comment by afbjorklund** [comment]
BTW, the docker socket is available in ~/.lima/docker/sock/docker.sock

`brew install lima`

`limactl start template://docker`

`docker.lima version`

I'm not sure how the "lima" extension is supposed to choose which engine

**Comment by slemeur** [comment]
Thanks for the report @afbjorklund and the description of the workaround - as you mentioned in https://github.com/containers/podman-desktop/issues/1081#issuecomment-1553288906 seems to be a regression. 

Would you be interested to contribute? We would actually really like to work with you on improving the support of Lima.

**Comment by afbjorklund** [comment]
Sure, I expect that you want Lima to be some kind of alternative to the default podman machine ?

It is possible to replace both the Podman and the Docker options, with a Lima instance running it

(the [Docker template](https://github.com/lima-vm/lima/blob/master/examples/docker.yaml) will run Ubuntu, the [Podman template](https://github.com/lima-vm/lima/blob/master/examples/podman.yaml) will run Fedora - but it can be tweaked)

As you know, containerd doesn't really offer a docker socket out-of-the-box - just hacks like "[nerdctld](https://github.com/afbjorklund/nerdctld#readme)"*

----

\* there are some experiments available, if adventurous: <https://github.com/afbjorklund/nerdctld/tree/main/lima>

`~/.lima/nerdctl/sock/nerdctl.sock`

Something like: `limactl start https://raw.githubusercontent.com/afbjorklund/nerdctld/main/lima/nerdctl.yaml`

It's just a proxy, to the `nerdctl` CLI.

**Comment by afbjorklund** [comment]
We talked about it earlier (fixing lima in PD), but it didn't go anywhere...
https://kubernetes.slack.com/archives/C04A0L7LUFM/p1668433204450239

Currently it seems broken, it just shows status "unknown" but no containers/images ?
Not sure how the socket location gets communicated with the podman / docker code.

```
$ podman-remote --url unix://$HOME/.lima/podman-lima/sock/podman.sock images
REPOSITORY                 TAG         IMAGE ID      CREATED     SIZE
docker.io/library/busybox  latest      af2c3e96bcf1  6 days ago  5.09 MB
localhost/limaimage        latest      af2c3e96bcf1  6 days ago  5.09 MB
```

But nothing shown for "limaimage" filter under images, in Podman Desktop.
Same for "limacontainer" filer and containers, if starting one with podman.lima

**Comment by afbjorklund** [comment]
It was due to the status being left in 'unknown' forever, and never started.

Hacking it makes it show both the containers and images, as expected...

![pd-lima-containers](https://github.com/containers/podman-desktop/assets/10364051/a6182561-43f5-4702-95de-bcce9ec81cff)
![pd-lima-images](https://github.com/containers/podman-desktop/assets/10364051/f86ea09b-c6bb-4837-ada5-5af3999dc5f4)

```diff
index ab88f12d..ad6a5cdb 100644
--- a/extensions/lima/src/extension.ts
+++ b/extensions/lima/src/extension.ts
@@ -29,7 +29,7 @@ function registerProvider(
   const containerProviderConnection: extensionApi.ContainerProviderConnection = {
     name: 'Lima',
     type: 'podman',
-    status: () => 'unknown',
+    status: () => 'started',
     endpoint: {
       socketPath: providerSocketPath,
     },
```

It just seems unimplemented (the lima extension), and premature to include

Choosing between podman and docker should probably be configurable ?

**Comment by afbjorklund** [comment]
Currently nerdctld isn't working, since Podman Desktop doesn't negotiate an API version.

Docker client: `GET      "/v1.40/images/json"`

Podman desktop: `GET      "/images/json?all=false"`

Also the "system df" command is not implemented yet, but that's probably not as critical.

```console
$ nerdctl system df
FATA[0000] unknown subcommand "df" for "system"         
```

https://docs.docker.com/engine/api/v1.40/#tag/System/operation/SystemDataUsage

**Comment by benoitf** [comment]
@afbjorklund we could provide some configuration option in the settings like an enum property with a default value.  Then people can choose which provider they want using Lima

**Comment by afbjorklund** [comment]
I think it will need some of the logic from the podman driver, so that it can stop and start the VM instance as well.
But the documentation is somewhat misleading, since it says that Lima is a container runtime (and not containerd)

"Podman Desktop can control various container engines, such as:
* Docker
* Lima:
  - nerdctl (containerd/buildkitd)
  - Docker
  - Podman
* Podman"

Might be a better way to document, not sure how PD runs a Docker VM today ? Just install the Docker Desktop too ?
The start and stop button don't seem to be doing much, with regards to the Docker VM. It just keeps on running

EDIT: Starting and stopping the Lima VM is a separate feature, and it is out-of-scope for this particular issue...

**Comment by afbjorklund** [comment]
Need to override the "default" template for limactl, which is containerd:

`podman machine init --now` => `limactl start template://podman`

`podman machine start` => `limactl start podman`
`podman machine stop` => `limactl stop podman`

`podman machine rm -f` => `limactl delete podman`

----

Also needs some translation of the machine parameters, like cpu/memory.

```
 limactl start --set='.cpus = 2 | .memory = "1048MiB"'
```

* `podman.factory.machine.cpus`

* `podman.factory.machine.memory`

**Comment by afbjorklund** [comment]
### nerdctl

Testing nerdctld, with a temporary override of the provider and socket:

```diff
--- a/extensions/lima/src/extension.ts
+++ b/extensions/lima/src/extension.ts
@@ -28,8 +28,8 @@ function registerProvider(
 ): void {
   const containerProviderConnection: extensionApi.ContainerProviderConnection = {
     name: 'Lima',
-    type: 'podman',
-    status: () => 'unknown',
+    type: 'docker',
+    status: () => 'started',
     endpoint: {
       socketPath: providerSocketPath,
     },
@@ -42,7 +42,7 @@ function registerProvider(
 }
 
 export async function activate(extensionContext: extensionApi.ExtensionContext): Promise<void> {
-  const socketPath = path.resolve(os.homedir(), '.lima/podman-lima/sock/podman.sock');
+  const socketPath = path.resolve(os.homedir(), '.lima/nerdctl/sock/nerdctl.sock');
   const socketAlternativePath = path.resolve(os.homedir(), '.lima/podman-lima/sock/podman.sock');
 
   let provider;

```

Added basic support for system df, and for volumes, so that commands are happy.

1) it will just return zero for everything

```
TYPE            TOTAL     ACTIVE    SIZE      RECLAIMABLE
Images          0         0         0B        0B
Containers      0         0         0B        0B
Local Volumes   0         0         0B        0B
Build Cache     0         0         0B        0B
```

2) if failing to set api version, default it

```
// Apparently HTTP 302 doesn't work over unix:// sockets.
// Reroute the call, after adding a proper /1.40 version.
```

But there are still more requirements, before containers and volumes will show up...

1) ~~maybe /events ?~~

It was crashing if container Mounts or diskusage Volumes was null instead of empty

`TypeError: Cannot read properties of null (reading 'find')`

Not crashing, even though it just shows 0 seconds (for created) and 0 bytes (for size).

![pd-lima-volumes](https://github.com/containers/podman-desktop/assets/10364051/b47ffbde-17ff-4fba-93b8-ea07aa3a2ffa)

2) docker version

Here is the engine backend info, as reported over the nerdctl.sock

`docker --host unix://$HOME/.lima/nerdctl/sock/nerdctl.sock version`

```
Server: 🤓
 nerdctl:
  Version:          1.3.1
 buildctl:
  Version:          0.11.5
  GitCommit:        252ae63bcf2a9b62777add4838df5a257b86e991
 containerd:
  Version:          1.7.0
  GitCommit:        1fbd70374134b891f97ce19c70b6e50c7b9f4e0d
 runc:
  Version:          1.1.5
  GitCommit:        f19387a6
 tini:
  Version:          0.19.0
  GitCommit:        de40ad0
 Engine:
  Version:          1.3.1
  API version:      1.40 (minimum version 1.24)
  Go version:       go1.20.3
  Git commit:       b224b280ff3086516763c7335fc0e0997aca617a
  Built:            
  OS/Arch:          linux/amd64
  Experimental:     true
```

**Comment by afbjorklund** [comment]
@benoitf : supporting more than one engine can be a separate feature. it can hardcode "podman" for now...


template | name | engine | type | socket | client | notes
--- | --- | --- | --- | --- | --- | ---
default.yaml | default | default* | _\<none\>_ || `nerdctl.lima`
docker.yaml | docker | docker | docker | `docker.sock` | `docker.lima` | OS: ubuntu
podman.yaml | podman | podman | podman | `podman.sock` | `podman.lima`| OS: fedora
apptainer.yaml | apptainer | apptainer | _\<none\>_ | | `apptainer.lima`
[nerdctl.yaml](https://github.com/afbjorklund/nerdctld/blob/main/lima/nerdctl.yaml) | nerdctl | nerdctl* | docker | `nerdctl.sock` | `nerdctl.lima`<br>`docker.lima` | _limited and_<br>_experimental_

\* containerd/buildkitd

Note: The default instance does not **have** a remote API, so all the `lima nerdctl` commands are run over CLI.

The docker and podman instances will use a client if it exists, or fall back to run `lima docker` or `lima podman`.

**Comment by afbjorklund** [comment]
You can also run your Kubernetes (on k3s or k8s) with Lima, but I consider that to be out of scope for this issue... The kubeconfig from those instances should be easy enough to configure, like any other kube context. And VM support for Lima is not yet available anyway, for starting and stopping.

`limactl start template://k3s`
`limactl start template://k8s`
`limactl start template://nomad`
`limactl start template://faasd`

![lima-gui-quick-start](https://github.com/containers/podman-desktop/assets/10364051/289ccbf8-ebd2-494d-9ab9-a6b6f77a8607)

https://github.com/afbjorklund/lima-gui

**Comment by afbjorklund** [comment]
Now it works with the default installation. I guess the issue can be closed, fix in v1.0.1 ~~(or backported to 1.0.x)~~

Not sure if the version of Lima is all that interesting, similar to the version of Docker Desktop (i.e. not Docker) ?

----

Seemed to add a whole lot of copy/paste code, from the podman provider, for running a simple `--version`...

The link to the website also looked a bit odd, I guess people can find https://lima-vm.io/ without _too_ much effort ?

![pd-lima-version](https://github.com/containers/podman-desktop/assets/10364051/5a87775b-d495-49c0-b40f-ba52ee8561cb)

Better if it could show the container runtime details instead, like if running Podman or Docker and what version.

The default provider seem to be showing the **client** version, which can be slightly misleading (server is better)

```
Client:
Version:      3.4.2
API Version:  3.4.2
Go Version:   go1.15.2
Built:        Thu Jan  1 01:00:00 1970
OS/Arch:      linux/amd64

Server:
Version:      4.5.0
API Version:  4.5.0
Go Version:   go1.20.2
Built:        Fri Apr 14 17:42:22 2023
OS/Arch:      linux/amd64

```
